File: exercises/js_challenges/CtCI/chatper_2_linked_lists/2.1_remove_dups.js
// Write code to remove duplicates form an unsorted linked list.

// create a linked list 

function Node(data, next = null) {
  this.data = data;
  this.next = next;
}

let list = new Node(5)
list = new Node(4, list)
list = new Node(3, list)
list = new Node(10, list)
list = new Node(3, list)

// keep track of whats been visited. 
// and use a temperary buffer to swap

function remove_next(node) {
  node.next = node.next.next
  return node
}
function remove_dups(head) {
  let hash = {}
  temp = head
  hash[temp.data] = true
  while (temp.next != null) {
    if (hash[temp.next.data] === undefined) {
      hash[temp.next.data] = true
    } else {
      temp = remove_next(temp)
    }
    if(temp.next != null){
      temp = temp.next
    }
  }
  return head
}

remove_dups(list)
console.log(list)

File: exercises/ruby_challenges/CtCI/chapter_2_linked_lists/2.1_remove_dups.rb
# Write code to remove duplicates form an unsorted linked list.


# creating a linked list
class Node
  attr_accessor :data, :next
  def initialize(data, next_node = nil)
    @data = data
    @next = next_node
  end
end

list = Node.new(3)
list = Node.new(4,list)
list = Node.new(5,list)
list = Node.new(10,list)
list = Node.new(3,list)

def remove_next(node)
  node.next = node.next.next
  return node 
end 

def remove_dups(head)
  hash = {}
  temp = head
  hash[temp.data] = true
  while(temp.next != nil)
    if(hash[temp.next.data] === nil)
      hash[temp.next.data] = true 
      temp = temp.next
    else 
      temp = remove_next(temp)
      temp = temp.next 
    end
  end
  return head 
end

p remove_dups(list)
p list